domain/macaroon/state/rootkeys.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"time"
+
+	"github.com/canonical/sqlair"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery/dbrootkeystore"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	macaroonerrors "github.com/juju/juju/domain/macaroon/errors"
+	"github.com/juju/juju/internal/database"
+	internaldatabase "github.com/juju/juju/internal/database"
+)
+
+// RootKeyState describes the persistence layer for
+// macaroon root keys
+type RootKeyState struct {
+	*domain.StateBase
+}
+
+// NewRootKeyState return a new macaroon root key state reference
+func NewRootKeyState(factory coredatabase.TxnRunnerFactory) *RootKeyState {
+	return &RootKeyState{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetKey gets the key with a given id from dqlite. If not key is found, a
+// macaroonerrors.KeyNotFound error is returned.
+func (st *RootKeyState) GetKey(ctx context.Context, id []byte) (dbrootkeystore.RootKey, error) {
+	db, err := st.DB()
+	if err != nil {
+		return dbrootkeystore.RootKey{}, errors.Trace(err)
+	}
+	key := RootKey{
+		Id: id,
+	}
+	getKeyStmt, err := st.Prepare("SELECT &RootKey.* FROM macaroon_root_key WHERE id = $RootKey.id", key)
+	if err != nil {
+		return dbrootkeystore.RootKey{}, errors.Annotate(err, "preparing get root key statement")
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, getKeyStmt, key).Get(&key)
+		if internaldatabase.IsErrNotFound(err) {
+			return errors.Annotatef(macaroonerrors.KeyNotFound, "key with id %s", string(id))
+		}
+		return domain.CoerceError(err)
+	})
+	return domainRootKeyToBakeryRootKey(key), errors.Trace(err)
+}
+
+// FindLatestKey returns the most recently created root key k following all
+// the conditions:
+//
+// k.Created >= createdAfter
+// k.Expires >= expiresAfter
+// k.Expires <= expiresBefore
+//
+// If no such key was found, return a macaroonerrors.KeyNotFound error
+func (st *RootKeyState) FindLatestKey(ctx context.Context, createdAfter, expiresAfter, expiresBefore time.Time) (dbrootkeystore.RootKey, error) {
+	db, err := st.DB()
+	if err != nil {
+		return dbrootkeystore.RootKey{}, errors.Trace(err)
+	}
+
+	var key RootKey
+	m := sqlair.M{
+		"created_after":  createdAfter,
+		"expires_after":  expiresAfter,
+		"expires_before": expiresBefore,
+	}
+
+	q := `
+SELECT &RootKey.*
+FROM   macaroon_root_key
+WHERE  created_at >= $M.created_after
+AND    expires_at >= $M.expires_after
+AND    expires_at <= $M.expires_before
```

Comment:
Sqlite does not implement a [TIMESTAMP](https://www.sqlite.org/datatype3.html#date_and_time_datatype) column, so it's by fluke this works.

When we move to strict tables, this will be TEXT or INTEGER. If it becomes TEXT I don't believe this will work correctly.

Reply:
The lease domain uses `TIMESTAMP` and does LT comparisons with `datetime(now)`. Secrets also uses `DATETIME` fields.

Reply:
That maybe so, because the library correctly handles this, but my point was around strict tables.

Reply:
Discussed this further over a call

Our conclusion was that no change is needed at the moment